### DISCUSSION TITLE
Accessing nested objects

### DISCUSSION THREAD

**Discussion by v-razov** [nlohmann/json#2271] [Q&A] [answered]
Can you access nested objects using "at()" for example;
 
std::string sVal = j.at("a.b.c").get<std::string>();

I want to avoid j["a"]["b"]["c"].get<std::string>(); because i need my code to be configurable by a user and so want to avoid recursive loop etc. when the user enters "a.b.c"...

**Comment by nlohmann** [comment, marked as answer]
No, such nesting would have to be done with `j.at("a").at("b").at("c")`. Instead, you may try JSON Pointers (see https://github.com/nlohmann/json#json-pointer-and-json-patch) with syntax like `j.at("/a/b/c")`.

**Comment by v-razov** [comment]
That doesn’t seem to work. I get an exception.

**Reply by nlohmann** [reply]
What kind of exception?

In order to asses your issue, we need the following information:

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?

- What is the expected behavior?

- And what is the actual behavior instead?

**Comment by v-razov** [comment]
I get an error 403. This is a snippet;

std::string sTxn = "{\"reason\":{\"reasonCode\":\"2204\"}}";
json j = json::parse(sTxn);
std::string sValue = j.at("reason/reasonCode").get<std::string>();

**Reply by nlohmann** [reply]
You need to pass a JSON Pointer, and the pointer needs to start with `/`. See https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_ac07f0518ba7a3af394d37d86753a9845.html#ac07f0518ba7a3af394d37d86753a9845 for examples.

```cpp
#include <iostream>
#include "json.hpp"

using json = nlohmann::json;

int main()
{
    std::string sTxn = R"({"reason":{"reasonCode":"2204"}})";
    json j = json::parse(sTxn);
    std::string sValue = j.at("/reason/reasonCode"_json_pointer).get<std::string>();
    std::cout << sValue << std::endl;
}
```

Output: `2204`

**Comment by v-razov** [comment]
That works! Thank you 👍

**Comment by v-razov** [comment]
Can i pass a variable to at() that contains the nested objects i want? i.e. std::string sVar = "/reason/reasonCode";

**Reply by nlohmann** [reply]
Yes, you need to create a json_pointer object from it.

**Comment by v-razov** [comment]
Can i pass a string variable to a json_pointer somehow? The idea is to have the user input what nested object they want to extract from the JSON payload.

I cant seem to find any examples from the above link.

**Reply by nlohmann** [reply]
Yes.

**Comment by v-razov** [comment]
Do you have an example?

**Reply by nlohmann** [reply]
See the docs on class json_pointer

**Comment by v-razov** [comment]
Have it working now. Thanks for the help much appreciated.
